### PR TITLE
Wikidata Lexeme Connector

### LINKED ISSUE #8: Support for Wikidata Lexemes
It is very useful to be able to search for Q-Items in Wikidata, but Wikidata also provides a database of Lexemes which could be used for annotating e.g. part of speech tag information or information about the word itself or even deriving those from the words morphology
https://www.wikidata.org/wiki/Wikidata:Lexicographical_data
Are there any plans on incorporating lexemes into this plugin or would one just write a new connector for that?

### PR DESCRIPTION
**Pull request by situx**
This pull request closes #8 and adds a Wikidata Lexeme Connector.
The Lexeme connector has two functions:

1.  Mark a text, find its lexeme, and add its lexeme URI and item sense URI to the widget
2.  Mark a text, find its lexeme, and add its part of speech tag URI to the widget

I have set up a test page in my fork, so you can try marking some french words https://situx.github.io/recogito-semantic-tags/

Apart from the connector itself, I changed the following things in the code:

- Update dependencies to be able to use the onBatchModify function to add/update two tags at once
- Modify the suggestion-loaded class to show wordforms and senses of lexemes if available
- Modify the onUpsertTag function to look for more than one tag to add/update

Will be happy about any feedback or improvements from your side and would love to see this connector added by default in the plugin.

### CODE REVIEW

File: public/index.html [hunk 1]
```diff
@@ -693,6 +693,8 @@ <h4>Arithm&eacute;tique et g&eacute;om&eacute;tries th&eacute;oriques et pratiqu
           // customization
           dataSources: [
             'Wikidata',
+			'WikidataLexeme',
+            { source: 'WikidataLexeme', name: 'WD Lexeme Wordform', wordformmode:true },
```
Comment (jamiefolsom) [inline public/index.html]: New connectors like this one should be optional and configurable; @rsimon does this comport with your vision for connectors?

Comment (jamiefolsom) [reply public/index.html]: Also, we may want to provide some guidance for implementers as to how to label services, since if the list grows a lot and contains lots of long labels, inconsistent labels, or abbreviations, it could get tricky for users to find what they want, given the small side-scrolling menu space.

File: src/index.jsx [hunk 2]
```diff
@@ -88,10 +88,24 @@ const SemanticTagPlugin = config => props => {
...
   const onUpsertTag = tag => {
     const existing = tagBodies.find(b => b.source === tag.uri);
-    if (existing)
-      props.onUpdateBody(existing, tagToBody(tag));
-    else
-      props.onAppendBody(tagToBody(tag));
+	if(tag.senseuri){
+		const existing2 = tagBodies.find(b => b.source === tag.uri2);
+		if(existing && existing2){
+			props.onBatchModify([{"action":"update", "previous":existing,"updated":tagToBody(tag)},{"action":"update", "previous":existing2,"updated":tagToBody({uri:tag.uri2,label:tag.label2,description:tag.description2})}])
```
Comment (jamiefolsom) [inline src/index.jsx:94]: @rsimon Should we add style conventions or linting to a CONTRIBUTORS.md file? Line length is an issue in this case, but also maybe unwieldy if/else.

File: src/connectors/WikidataLexeme.js [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+	  ORDER BY ?l ?lfgram ?sense ?senselabel 
+      LIMIT ${limit}
+    `;
+
+    const url = wd.sparqlQuery(sparql);
+    return fetch(url)
+      .then(response => response.json())
+      .then(data => data.results.bindings.map(result => {
+        return { 
+          uri: result.lf?.value,
```
Comment (jamiefolsom) [inline src/connectors/WikidataLexeme.js:120]: @situx it might be nice to factor this out into a more easily configurable format; at the moment, we don't have a convention for that, so either we can provide one, or make a change in a future PR.

File: package.json [hunk 4]
```diff
@@ -10,31 +10,31 @@
...
-    "webpack": "^5.66.0",
-    "webpack-cli": "^4.9.1",
+    "webpack": "^5.74.0",
+    "webpack-cli": "^4.10.0",
     "webpack-dev-server": "^4.7.3"
   },
   "peerDependencies": {
-    "react": "^16.14.0",
-    "react-dom": "^16.14.0"
+    "react": "^18.2.0",
```
Comment (jamiefolsom) [inline package.json:28]: Any update to versions should be cleared with @rsimon and well-tested. Are these updates necessitated by the addition of the WikiData Lexeme connector? If not, they should be factored out into another PR, so that testing doesn't block merge of this PR.

**Review by jamiefolsom — CHANGES_REQUESTED** [review]
@situx This is fascinating, and likely would have wide appeal, so first off, thank you for the PR! 

I would like (and have requested) to have @rsimon review this, as he's the main developer. He's out of the office this week, so it may be a few days or a week before he can chime in.

Couple of items I would like to see adjusted, if you're willing.

- If the version updates in package.json are not necessary to this functionality, let's factor them out into a separate PR. I'm all for keeping up with versions, but just separately from the core feature, so they don't block your core contribution.
- I notice that upon selection of a suggested match, two tags are added to the field. Can you explain what those are, and why there are two? The user can remove one they don't want, but it's not readily evident what they are. 
- Also, typically, we don't add two tags with one click -- it would be ideal to allow the user to select which of the results to add, perhaps by returning more than one result, if more than one exists.
- There are a few code style nits here, but we haven't set standards, so either we'll do that, and request a few adjustments, or let them slide for now, and address them in a future PR.
- Would you be willing to add a block to the README, explaining a bit more about this connector? I understand it leverages WD's lexeme service, but my college linguistics courses are in the somewhat distant past at this point, so it might be useful to me and others if you could enlighten us somewhat.
- We want this to be easily configurable, and optional, so I will let @rsimon chime in on how best to do that, in this case, and in future contributions.

Thanks again!

File: package.json (hunk 4 above)
Comment (situx) [reply package.json:28]: Sure thing. I will test and reduce the number of dependencies to the bare minimum. I updated more than needed

**Comment by situx** [comment]
Hi, thanks for your feedback, and yes, I am all for sorting code styles and other things out before merging.

- Dependencies: Actually, the only dependency that needs updating is recogito-client-core, but when building with npm I realized that this requires babel/core7, so that also seems to be in need of updating. While at it, I also updated other dependencies, but I could possibly reduce it to recogito-client-core and babel/core7 if you like

Then you were asking for some more information about the connector, and I will answer with an example:

Imagine you mark the word "dogs" in your text, then the connector enables you to find the Lexeme for "dog" in Wikidata (https://www.wikidata.org/wiki/Lexeme:L1122) via its wordform http://www.wikidata.org/entity/L1122-F2 "dogs" i.e., dog in the plural. 

However, I find the Lexeme URI not necessarily to be the only item that you might want to add to your annotation. Maybe you want to add one of its two meanings/senses. (dog the animal (https://www.wikidata.org/wiki/Q144) or dog the tool (https://www.wikidata.org/wiki/Q5288153))
If so, then this would be the default functionality of the connector: Adding the URI for the Lexeme and adding the URI for one of its meanings (the one you select in the dropdown)

Or you might want to add part of speech tags. If the connector is initialized with  wordformmode:true, then instead of adding sense URIs it will add URIs for the word form, in the case of "dogs" plural (https://www.wikidata.org/wiki/Q146786)

Now that I write this down, maybe you would also like to add the wordform URI and the sense URI without adding the Lexeme URI, or even all three. I had no need for this particular use case, but maybe it is interesting for some people.

So in my use case, I am annotating words in cuneiform transliterations in the Latin script. Depending on the particular case, I would like to retain a link to the word form and its meaning in the text or the part of speech tags associated with the word.

From the perspective of usability, I found it is way more convenient just to mark a word and let Wikidata find the list of meanings for you. Before, either you were familiar with the language you annotate, and you needed to search for the concept using the Wikidata connector (doable but compared to "just marking" tedious), or you were not familiar with the language you annotate, and you need to look up the meaning of the word somewhere first before using the Wikidata connector to annotate the meaning. Both use cases are likely to become much easier with a Lexeme connector.

File: src/connectors/WikidataLexeme.js (hunk 3 above)
Comment (situx) [reply src/connectors/WikidataLexeme.js:120]: Yes, I think that would be a good idea. Would be nice to get some guidance here

**Comment by jamiefolsom** [comment]
I see that I overlooked some of the answers to my questions in your cover letter. This is a nice, clear explanation of the use case(s). Would you include that in an addendum to the readme?

It's fine to include updates to dependencies that are relevant to this PR, of course. It sounds like Rainer needs to respond to the mismatch in requirements between recogito-client-core and recogito-semantic-tags.

While we're waiting for his review, it might be nice to rebase/squash commits, so we can keep the commit history nice and neat. Thanks again -- looking forward to using this myself!

**Comment by situx** [comment]
While testing the connector on the cuneiform transliteration texts today, I found the need to add another function to the plugin:
In cuneiform transliteration texts, there are some additional characters which are added to a word, such as # indicating that the particular sign is broken.
In a dictionary, these additional characters will never appear. Hence they would need to be filtered before querying their meaning.
To that end, two new parameters can be given to the connector now:
``
{ source: 'WikidataLexeme', name: 'WD Lexeme Regex', wordformmode:false, normpattern:/[\[\]\(\),\.#!\?\}\{]/g, targetpattern:""},
``
_normpattern_ defines a regular expression or string of characters to be filtered. By default, these are replaced by the empty string, but if for whatever reason you want them to be replaced with something else, you could use the optional _targetpattern_ parameter.
On the test page (https://situx.github.io/recogito-semantic-tags/), I have added some test cases of mine for this in Sumerian (choose language sux):
For example: "lugal?" means the author is unsure whether this is the sign lugal. Without regex, the connector will not find any results; with regex, it will find results for lugal.
It might apply to other languages as well, if only for convenience reasons. Maybe you annotate the last word of a question, including the question mark, and still want to find results for the last word.

**Comment by situx** [comment]
Finally, I extended _normpattern_ also to accept a callback. This is necessary because there are different variants of writing cuneiform transliterations.
For example, the expressions uš11 and usz11 are equivalent, but only usz11 might be given in the dictionary. The transformations of different transliteration styles can get so complex that they cannot be expressed in a single regular expression. As this is my first time implementing a callback, maybe you have some suggestions, but as always, you can test it on the example page. https://situx.github.io/recogito-semantic-tags/

**Comment by jamiefolsom** [comment]
@situx I like this, and will look forward to hearing @rsimon's thoughts when he returns from holiday.

File: package.json (hunk 4 above)
Comment (rsimon) [reply package.json:28]: Minor version upgrades are fine, although I agree they should be in a separate PR. Concerning React: given that this is a peer dependency, this probably doesn't matter too much. But I've been testing compatibility only with React 17 so far. (17.0.2, to be specific). I'd leave it at this version for now.

@situx can I ask what your current build setup is? Are you using the plugin in your own React host app? Or compiling just the plugin bundle and then using it in a non-React app?

File: public/index.html (hunk 1 above)
Comment (rsimon) [reply public/index.html]: Yes, that's how it should be. I.e. The config would add another datasource, just like above.

File: src/components/SuggestionsLoaded.jsx [hunk 5]
```diff
@@ -32,16 +32,20 @@ const SuggestionsLoaded = props => {
                   onClick={evt => evt.stopPropagation()}
                   href={suggestion.uri}
                   target="_blank">
-                  <BiLinkExternal className="external"/>
+                  <BiLinkExternal className="external"/> 
                 </a>
+				{suggestion.wordformuri ? <a href={suggestion.wordformuri} className="id" onClick={evt => evt.stopPropagation()} target="_blank">({suggestion.wordformlabel})</a>:''}
+				{suggestion.senseuri ? <a href={suggestion.senseuri} className="id" onClick={evt => evt.stopPropagation()} target="_blank">[{suggestion.senselabel}]</a>:''}
               </> :
```
Comment (rsimon) [inline src/components/SuggestionsLoaded.jsx:39]: Hm, I really need to think about this one. `SuggestionsLoaded` should remain a generic component. Having a hard-wired dependency to the data structure that's produced by one specific data source is not good. This component shouldn't have to be concerned about whether the suggestion has a "wordformuri". (Also just to be nitpicky: can we camel-case that property to `wordFormUri`?)

File: src/connectors/WikidataLexeme.js [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+				target=this.config?.targetpattern
+			}
+			try{
+				query=query.replaceAll(this.config?.normpattern,target)
+			}catch(error){
+				console.log(error)
+			}		
+		}
+	}
+	console.log(query)
```
Comment (rsimon) [inline src/connectors/WikidataLexeme.js:70]: I think we can remove this debug log?

File: src/index.jsx (hunk 2 above)
Comment (rsimon) [reply src/index.jsx:94]: Likewise, I think this needs to be rethought. It's not good if the `index.jsx` includes hard-wired dependencies on the data structure (`tag.senseuri`) of one specific connector.

I need more time to look into this. But I think instead, `index.jsx` probably needs some kind of `onBatchModify` method, while all the logic that's now packed into this if-block needs to reside somewhere else, in the connector implementation. Otherwise we just scatter logic related to handling the Lexeme-Tags all over the plugin.

**Review by rsimon — CHANGES_REQUESTED** [review]
Thanks for this PR - it will potentially add a useful extension to the plugin. 

However at the moment, I think this still needs a bit of refactoring. My main issue (as written in the inline comments below) is that logic specific to the Lexeme Connector is scattered in other regions of the plugin, which really should not have any hard-wired dependencies to one specific connector. This will make things harder to maintain in the future.

I can see *why* those changes were made. But in both cases (`index.js` and `SuggestionsLoaded.jsx`) we need to find changes that make the plugin more generic, so that it can facilitate the kind of operations the LexemeConnector needs (=adding multiple tags), without polluting the rest of the plugin with LexemeConnector specifics.

Also, we should perhaps talk about the data model of the query response. I admit I could not go in depth right now, due to lack of time. But it might be possible to make the data structure a bit more readable if we remodel it a bit (which then might also inform the changes we need to make to `index` and `SuggestionsLoaded`).

**Comment by rsimon** [comment]
+1 on the features to pre-filter the annotated string.

Would it be possible to add markdown documentation of the plugin functionality and available config props? My suggestion would be to add it in a GitHub issue. And then I can transfer it to the Wiki, where the other connectors are documented.

File: src/components/SuggestionsLoaded.jsx (hunk 5 above)
Comment (situx) [reply src/components/SuggestionsLoaded.jsx:39]: I am just compiling the bundle and using it in a non-React app in the way that you see on my test homepage (linked above).
As for SuggestionsLoaded, surely you could name it uri2 and uri3 or sth. like that, but I agree, that I cannot provide an optimal solution there. That would be up to you as lead developers to decide.
Indeed it is probably the first time you would want to add two URIs at once or display information contained in 3 different URIs.
In any case, happy to add any changes and the branch is also open for you as maintainers of this repository to change anything you might want to change

File: src/connectors/WikidataLexeme.js (hunk 6 above)
Comment (situx) [reply src/connectors/WikidataLexeme.js:70]: Correct

**Comment by situx** [comment]
Yes, I will write some documentation about the connectors options and functionalities

**Comment by situx** [comment]
I came across another interesting case I would like to cover:
image

Here, one part of the annotated text is in superscript `<sup></sup>`, so recogito will filter the superscript tags (like it normally should), and then my search will yield no results because the words have been merged.
In linked data dictionaries, likely, superscript will not be supported simply because it is a markup feature and not part of the Unicode specification (except for some exceptions).
In my case, the word gambimušen will be present as gambi{mušen} in the dictionary, with {mušen} indicating that mušen is written in superscript.
Long story short, I would like to ask if there is a way to access the HTML version of whatever is marked with recogito, so that I could replace e.g. `<sup>` with { and `</sup>` with } before sending my query.
@rsimon @jamiefolsom any ideas?

**Comment by situx** [comment]
I added an ugly workaround for now, but it works:
image
* I could not access the selection from the web browser, because once the Semantic Tags Window is open, the selection in the browser becomes a special `<div>` tag with class `r6o-selection`
* So I did the next best thing: Iterate over all `<div class="r6o-selection">` and search its parents for superscript/subscript tags
* Reconstruct the HTML and then conduct replacements

Would be much easier if recogito could return a HTML tag of what is annotated in HTML though
